arrow/obsidian.c

```diff
@@ -0,0 +1,279 @@
+      interval = -log(1 - sample) / total;
+      point = ((double) rand() / RAND_MAX) * total;
+
+      // Based on the random sample, find the event that it corresponds to by
+      // iterating through the propensities until we surpass our sampled value
+      choice = 0;
+      progress = 0.0;
+      while(progress + propensities[choice] < point) {
+        progress += propensities[choice];
+        choice += 1;
+      }
+
+      // If we have surpassed the provided duration we can exit now
+      if (choice == -1 || (now + interval) > duration) {
+        break;
+      }
+
+      // Increase time by the interval sampled above
+      now += interval;
+
+      // Record the information about the chosen event this step
+      time[step] = now;
+      events[step] = choice;
+
+      // For each substrate involved in this reaction, update the ongoing state
+      // with its value from the stoichiometric matrix
+      substrates_length = substrates_lengths[choice];
+      for (involve = 0; involve < substrates_length; involve++) {
+        index = substrates_indexes[choice] + involve;
+        adjustment = stoichiometry[choice * substrates_count + substrates[index]];
+        outcome[substrates[index]] += adjustment;
+      }
+
+      // Find which propensities depend on this reaction and therefore need to be
+      // updated in the next step (the rest of the propensities will not change)
+      update_length = dependencies_lengths[choice];
+      for (up = 0; up < update_length; up++) {
+        index = dependencies_indexes[choice] + up;
+        update[up] = dependencies[index];
+      }
+
+      // Advance
+      step += 1;
+
+      // If our step has advanced beyond the current `event_bounds`, double the
+      // `event_bounds` and reallocate these arrays with the new size
+      if (step >= event_bounds) {
+        double * new_time = malloc((sizeof (double *)) * event_bounds * 2);
+        memcpy(new_time, time, (sizeof (double *)) * event_bounds);
+        free(time);
+        time = new_time;
+
+        long * new_events = malloc((sizeof (long *)) * event_bounds * 2);
+        memcpy(new_events, events, (sizeof (long *)) * event_bounds);
+        free(events);
+        events = new_events;
+
+        event_bounds *= 2;
+      }
+    }
+  }
+
+  // Construct the `evolve_result` from the results of performing steps until either
+  // the desired duration was achieved or there are no more reactions to perform.
+  // We return:
+  //   * How many steps were performed
+  //   * An array of the time of each event
+  //   * An array of each event that occurred each time step
+  //   * The resulting state of the system after the chosen reactions were applied
+  evolve_result result = {step, time, events, outcome};
+
+  // Clean up the transient allocations
+  free(propensities);
+  free(update);
+
+  // Return the result constructed above
+  return result;
+}
+
+// Print an array of doubles
+int
+print_array(double * array, int length) {
```

Comment:
Yeah this was purely debug stuff. I'm keeping it around out of nostalgia and in case I ever need to modify this it is fairly useful (though there is probably some logging library I could use instead.... trying to keep it simple).